execute/group_lookup.go

```diff
@@ -7,102 +7,242 @@ import (
 )
 
 type GroupLookup struct {
-	groups groupEntries
+	// groups contains groups of group keys in sorted order.
+	// These are optimized for appending access.
+	groups []*keyGroup
 
-	//  range state
-	rangeIdx int
+	// lastIndex contains the last group that an entry was
+	// found in or appended to. This is used to optimize appending.
+	lastIndex int
 
-	// Indicates whether the group needs to be sorted.
-	needSort bool
+	// nextID is the next id that will be assigned to a key group.
+	nextID int
 }
 
-type groupEntry struct {
-	key   flux.GroupKey
-	value interface{}
+// keyGroup is a group of keys in sorted order.
```

Comment:
It would help to make this code easier to understand if you used a word besides `group` to describe a collection of group keys in sorted order... Not sure what the best word would be though.

Reply:
`keyList` maybe?

Reply:
I like this idea. I'm changing it to `groupKeyList`.